README.md
# MPMP - Assembler

## Installation
### Precompiled Binaries
Download the latest precompiled binaries for your system (Architecture + OS) from [Releases](https://gitlab.com/moseschmiedel/masm/-/releases) tab.

### Manual
#### Requirements
- rust
- cargo
#### Compilation
Clone this repository and checkout the desired release:
```sh
git clone https://gitlab.com/moseschmiedel/masm.git
cd masm
git checkout <desired-release-branch>
```
Compile and install the assembler locally via cargo:
```sh
cargo install --path .
```

## Usage
Using the assembler is straight-forward. It takes one input-assembly-file and produces a `.hex` output-file.
The name of the output-file can be specified.
```sh
$ masm --help
Usage: masm [OPTIONS] <INPUT_PATH>

Arguments:
  <INPUT_PATH>

Options:
  -o, --output <OUTPUT_PATH>  Output file where binary is stored
  -d, --debug                 Enable debug output to stdout
  -h, --help                  Print help
  -V, --version               Print version
```

## Author
Mose Schmiedel

## Copyright and License

This software is copyright (c) 2023 by Mose Schmiedel
masm is licensed under the [MIT](LICENSE.TXT) license.


CHANGELOG.md
# Changelog

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.4.0] - 2023-09-21
### Improved
 - allow underscores in label names

### Added
 - comments with semicolon character ';'


## [1.3.0] - 2023-06-26
### Improved
 - label parsing:
    - forward jump-labels are now possible
    - label definitions start at line start and end with colon ':'
    - label references don't need dot '.' anymore
 - whitespace handling:
    - all types of whitespace before instructions now get recognized
    - whitespace after label definition doesn't effect parsing result

## [1.2.0] - 2023-06-15
### Added
 - instructions:
    - Load `ld` and Store `st` from/to RAM
 - numbered register identifier (reg0..reg7)

### Removed
 - macOS continuous delivery build

## [1.1.0] - 2023-06-12
### Added
 - instructions:
    - `add3` - addition with 3 operands
    - `nop` - No operation
    - Absolute/relative jump
 - Jump to Label management

## [1.0.0] - 2023-06-05

### Added
 - instructions:
    - ALU
    - `hlt` - HALT
    - `mov` - MOVE register to register
    - `ldcon` - Load 16bit constant to register


test.s
    ldc %regF 0x7001
    ldc %regH 0xAAAA
    st %regF %regH
    ld %regB %regF


tests/test.s
    ldc %reg5 0x7001
    ldc %reg7 0xAAAA
    st %reg5 %reg7
    ld %reg1 %reg5


tests/comments.s
;;; this is the test file for comments

; this is a comment

ldc %reg0 8 ; this comment describes that this instruction loads 8 into the register 0



src/ir.rs
use std::{
    collections::HashMap,
    hash::{Hash, Hasher},
};

/// Main struct with the parser result. `IR` stands for
/// `intermediate representation` and consists of three components:
///     - `start_label` - reference to the label at which the execution should be started
///     - `label_definitions` - lookup-table for the labels, used for getting the address
///         of the LabelDefinition referenced by a LabelReference
///     - `instructions` - collection which stores for each label the associated instructions
///         in a `Vec`
pub struct IR {
    pub start_label: LabelReference,
    pub label_definitions: LabelLUT,
    pub instructions: HashMap<LabelReference, Vec<Instruction>>,
}

#[derive(Debug, Clone)]
pub struct LabelLUT(pub HashMap<LabelReference, LabelDefinition>);

impl LabelLUT {
    pub fn new() -> Self {
        LabelLUT(HashMap::new())
    }
    pub fn with_capacity(capacity: usize) -> Self {
        LabelLUT(HashMap::with_capacity(capacity))
    }
}

impl Default for LabelLUT {
    fn default() -> Self {
        LabelLUT::new()
    }
}

#[derive(Debug, Clone)]
pub struct LabelDefinition {
    pub name: String,
    pub address: MemoryAddress,
}

#[derive(Debug, Clone)]
pub struct LabelReference(String);

impl LabelDefinition {
    pub fn new(name: impl Into<String>, address: u16) -> LabelDefinition {
        LabelDefinition {
            name: name.into(),
            address: MemoryAddress(address),
        }
    }
}

impl LabelReference {
    pub fn new(name: impl Into<String>) -> Self {
        LabelReference(name.into())
    }
    pub fn name(&self) -> &str {
        &self.0
    }
}

impl From<LabelDefinition> for LabelReference {
    fn from(value: LabelDefinition) -> Self {
        LabelReference(value.name)
    }
}

impl Hash for LabelReference {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.name().hash(state);
    }
}

impl PartialEq for LabelReference {
    fn eq(&self, other: &Self) -> bool {
        self.name() == other.name()
    }
}

impl Eq for LabelReference {}

/// Enum which represents all possible instructions
/// and its metadata for the assembled language
#[derive(Debug, PartialEq, Eq)]
pub enum Instruction {
    Move(UnaryExpression),
    Load {
        address: RegisterAddress,
        source: LoadSource,
    },
    StoreRAM {
        address_register: RegisterAddress,
        data_register: RegisterAddress,
    },
    Halt,
    Noop,
    Jump {
        target: JumpTarget,
        condition: JumpCondition,
    },
    Add(BinaryExpression),
    Add3(TernaryExpression),
    AddWithCarry(BinaryExpression),
    Subtract(BinaryExpression),
    SubtractWithCarry(BinaryExpression),
    Increment(UnaryExpression),
    Decrement(UnaryExpression),
    Multiply(BinaryExpression),
    Test(BinaryStatement),
    AND(BinaryExpression),
    OR(BinaryExpression),
    NOT(UnaryExpression),
    XOR(BinaryExpression),
    XNOR(BinaryExpression),
    ShiftLeft(BinaryExpression),
    ShiftRight(BinaryExpression),
    Negate(UnaryExpression),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub struct RegisterAddress(pub u8);
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub struct MemoryAddress(pub u16);
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub struct Constant(pub u16);

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub struct Register {
    pub address: RegisterAddress,
}

impl Register {
    pub fn new(address: RegisterAddress) -> Self {
        Self { address }
    }

    pub fn addr(&self) -> u8 {
        self.address.0
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct UnaryExpression {
    pub target: Register,
    pub source_a: Register,
}

impl UnaryExpression {
    pub fn new(target: Register, source_a: Register) -> Self {
        Self { target, source_a }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct UnaryStatement {
    pub source_a: Register,
}

impl UnaryStatement {
    pub fn new(source_a: Register) -> Self {
        Self { source_a }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct BinaryExpression {
    pub target: Register,
    pub source_a: Register,
    pub source_b: Register,
}

impl BinaryExpression {
    pub fn new(target: Register, source_a: Register, source_b: Register) -> Self {
        Self {
            target,
            source_a,
            source_b,
        }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct BinaryStatement {
    pub source_a: Register,
    pub source_b: Register,
}

impl BinaryStatement {
    pub fn new(source_a: Register, source_b: Register) -> Self {
        Self { source_a, source_b }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TernaryExpression {
    pub target: Register,
    pub source_a: Register,
    pub source_b: Register,
    pub source_c: Register,
}

impl TernaryExpression {
    pub fn new(
        target: Register,
        source_a: Register,
        source_b: Register,
        source_c: Register,
    ) -> TernaryExpression {
        TernaryExpression {
            target,
            source_a,
            source_b,
            source_c,
        }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub enum LoadSource {
    Constant(u16),
    RAM { address_register: Register },
    Pgm,
}

#[derive(Debug, PartialEq, Eq)]
pub enum JumpTarget {
    Constant(u16),
    Register(Register),
    Label(LabelReference),
}

#[derive(Debug, PartialEq, Eq)]
pub enum JumpCondition {
    True,
    Zero,
    NotZero,
    Less,
}


src/main.rs
use std::{
    fs::File,
    io::{BufWriter, Write},
    process,
};

use clap::Parser;

use masm::{generator, lexer, parser};

#[derive(Parser)]
#[command(author, version, about, long_about=None)]
struct Cli {
    /// Output file where binary is stored
    #[arg(short, long = "output")]
    output_path: Option<std::path::PathBuf>,
    /// Enable debug output to stdout
    #[arg(short, long = "debug")]
    debug_enable: bool,

    input_path: std::path::PathBuf,
}

fn main() {
    let cli = Cli::parse();
    let input_path = cli.input_path.canonicalize().unwrap_or_else(|err| {
        eprintln!("Error: Could not find input file:");
        eprintln!("{err}");
        process::exit(1);
    });
    let output_path = cli.output_path.unwrap_or("output.hex".into());

    if cli.debug_enable {
        println!("Input: {}", input_path.display());
        println!("Output: {}", output_path.display());
    }

    let lexed = lexer::lexer(&input_path).unwrap_or_else(|errors| {
        for err in errors {
            eprintln!("Lexer: {err}");
        }
        process::exit(1);
    });

    let parsed = parser::parser(lexed).unwrap_or_else(|err| {
        eprintln!("Parser: {err}");
        process::exit(1);
    });

    if cli.debug_enable {
        println!("{:#?}", parsed.instructions.keys());
        println!("{:#?}", parsed.instructions.values());
    }

    let binary = generator::generator(parsed).unwrap_or_else(|err| {
        eprintln!("Generator: {err}");
        process::exit(1);
    });

    if cli.debug_enable {
        println!("{:#?}", binary);
    }
    let output = File::create(&output_path).unwrap_or_else(|err| {
        eprintln!("Error: Could not open output file for writing:");
        eprintln!("{err}");
        process::exit(1);
    });
    let mut writer = BufWriter::new(output);
    writer
        .write_all("v3.0 hex words plain\n".as_bytes())
        .and_then(|_| {
            for instr_line in binary.chunks(8) {
                let mut line = String::new();
                for instr_word in instr_line {
                    line = format!("{line} {instr_word}");
                }
                writer.write_all(format!("{}\n", line.trim()).as_bytes())?;
            }
            Ok(())
        })
        .and_then(|_| writer.flush())
        .unwrap_or_else(|err| {
            eprintln!("Error: Could not write to file:");
            eprintln!("{err}");
            process::exit(1);
        });
}


src/language.rs
pub enum Command {
    EmptyLine,
    Label {
        label: String,
        address: MemoryAddress,
    },
    Move(UnaryExpression),
    Load {
        source: LoadSource,
    },
    StoreRAM,
    Noop,
    Jump {
        condition: JumpCondition,
        negate: bool,
    },
    Add(BinaryExpression),
    AddWithCarry(BinaryExpression),
    Subtract(BinaryExpression),
    SubtractWithCarry(BinaryExpression),
    Increment(UnaryExpression),
    Decrement(UnaryExpression),
    Multiply(BinaryExpression),
    Test(BinaryStatement),
    AND(BinaryExpression),
    OR(BinaryExpression),
    NOT(UnaryExpression),
    XOR(BinaryExpression),
    XNOR(BinaryExpression),
    ShiftLeft(BinaryExpression),
    ShiftRight(BinaryExpression),
    Negate(UnaryExpression),
}

pub struct RegisterAddress(u8);
pub struct MemoryAddress(pub u16);

pub struct Register {
    adress: RegisterAddress,
}

pub struct UnaryExpression {
    target: Register,
    source_a: Register,
}

pub struct BinaryExpression {
    target: Register,
    source_a: Register,
    source_b: Register,
}

pub struct BinaryStatement {
    source_a: Register,
    source_b: Register,
}

struct TernaryExpression {
    target: Register,
    source_a: Register,
    source_b: Register,
    source_c: Register,
}

pub enum LoadSource {
    Constant(u16),
    RAM { address: MemoryAddress },
    Pgm,
}

pub enum JumpCondition {
    True,
    Zero,
    Less,
}


src/lexer.rs
use std::{
    collections::VecDeque,
    fs::File,
    io::{self, BufRead},
    path::Path,
};

pub trait LineNumber {
    fn get_line_number(&self) -> u16;
}

/// Keywords are the Tokens, that the lexer creates from the
/// input character stream
#[derive(Debug)]
pub enum Keyword {
    Mmenonic {
        name: String,
        line_number: u16,
    },
    RegisterAddress {
        name: String,
        line_number: u16,
    },
    Constant {
        value: u16,
        line_number: u16,
        origin: String,
    },
    Label {
        name: String,
        line_number: u16,
    },
}

impl Keyword {
    pub fn mmenonic(name: &str, line_number: u16) -> Keyword {
        Keyword::Mmenonic {
            name: name.to_string(),
            line_number,
        }
    }
    pub fn register_address(name: &str, line_number: u16) -> Keyword {
        Keyword::RegisterAddress {
            name: name.to_string(),
            line_number,
        }
    }
    pub fn constant(origin: &str, value: u16, line_number: u16) -> Keyword {
        Keyword::Constant {
            origin: origin.to_string(),
            value,
            line_number,
        }
    }
    pub fn label(name: &str, line_number: u16) -> Keyword {
        Keyword::Label {
            name: name.to_string(),
            line_number,
        }
    }
    pub fn get_original_string(&self) -> String {
        match &self {
            Keyword::Mmenonic { name, .. } => name.clone(),
            Keyword::RegisterAddress { name, .. } => format!("%{}", name.clone()),
            Keyword::Label { name, .. } => format!(".{}", name.clone()),
            Keyword::Constant { origin, .. } => origin.clone(),
        }
    }
}

impl PartialEq for Keyword {
    fn eq(&self, other: &Keyword) -> bool {
        match (self, other) {
            (
                Keyword::Label {
                    name: name_self, ..
                },
                Keyword::Label {
                    name: name_other, ..
                },
            ) => name_self == name_other,
            (
                Keyword::RegisterAddress {
                    name: name_self, ..
                },
                Keyword::RegisterAddress {
                    name: name_other, ..
                },
            ) => name_self == name_other,
            (
                Keyword::Mmenonic {
                    name: name_self, ..
                },
                Keyword::Mmenonic {
                    name: name_other, ..
                },
            ) => name_self == name_other,
            (
                Keyword::Constant {
                    value: value_self,
                    origin: origin_self,
                    ..
                },
                Keyword::Constant {
                    value: value_other,
                    origin: origin_other,
                    ..
                },
            ) => value_self == value_other && origin_self == origin_other,
            _ => false,
        }
    }
}

impl LineNumber for Keyword {
    fn get_line_number(&self) -> u16 {
        match *self {
            Keyword::Mmenonic { line_number, .. } => line_number,
            Keyword::RegisterAddress { line_number, .. } => line_number,
            Keyword::Constant { line_number, .. } => line_number,
            Keyword::Label { line_number, .. } => line_number,
        }
    }
}

///
#[derive(Debug)]
pub enum LexerError {
    InvalidRegisterIdentifier {
        actual: String,
        line_number: u16,
    },
    InvalidIdentifier {
        actual: String,
        line_number: u16,
    },
    CommandAfterCommand {
        command_name: String,
        line_number: u16,
    },
    LabelAfterCommand {
        label_name: String,
        line_number: u16,
    },
    IoError(io::Error),
}

impl std::fmt::Display for LexerError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self {
            LexerError::IoError(io_error) => write!(f, "IO error '{}'", io_error),
            LexerError::InvalidIdentifier {
                actual,
                line_number,
            } => write!(
                f,
                "Invalid identifier '{}' found at line {}",
                actual, line_number
            ),
            LexerError::LabelAfterCommand {
                label_name,
                line_number,
            } => write!(
                f,
                "Found illegal label '{}' after command at line {}",
                label_name, line_number
            ),
            LexerError::CommandAfterCommand {
                command_name,
                line_number,
            } => write!(
                f,
                "Found illegal command '{}' after command at line {}",
                command_name, line_number
            ),
            LexerError::InvalidRegisterIdentifier {
                actual,
                line_number,
            } => write!(
                f,
                "Invalid register identifier '{}' found at line {}",
                actual, line_number
            ),
        }
    }
}

/// The lexer reads the provided assembler text file and separate
/// it into Tokens (Keywords).
/// Tokens are strings that are separated by whitespace.
pub fn lexer(path: &Path) -> Result<Vec<Keyword>, Vec<LexerError>> {
    let mut errors: Vec<LexerError> = Vec::new();
    let file: File = File::open(path).map_err(|io_err| vec![LexerError::IoError(io_err)])?;
    let reader = io::BufReader::new(file);
    let mut line_number = 0;
    let mut lexed: Vec<Keyword> = Vec::with_capacity(32);
    let mut keyword_buffer: Vec<Keyword> = Vec::with_capacity(4);

    for line in reader.lines() {
        match line {
            Ok(line) => {
                match lex_line(&mut keyword_buffer, line, line_number) {
                    Ok(_) => lexed.append(&mut keyword_buffer),
                    Err(error) => errors.push(error),
                };
                line_number += 1;
            }
            Err(io_err) => {
                errors.push(LexerError::IoError(io_err));
                return Err(errors);
            }
        }
    }

    let hlt = Keyword::Mmenonic {
        name: String::from("hlt"),
        line_number,
    };
    if lexed
        .last()
        .and_then(|last_keyword| if *last_keyword != hlt { Some(()) } else { None })
        .is_some()
    {
        lexed.push(hlt);
    }

    Ok(lexed)
}

pub fn lex_line(
    keywords: &mut Vec<Keyword>,
    line: String,
    line_number: u16,
) -> Result<(), LexerError> {
    let mut line = line;
    // starts with 4 spaces -> instruction
    line = line.trim_end().to_string();
    if line.starts_with([' ', '\t']) {
        line = line.trim_start().to_string();
        if let Some(semi_idx) = line.find(';') {
            line.truncate(semi_idx);
        }
        let mut args: VecDeque<&str> = line.split_whitespace().collect();
        let command = args.pop_front().unwrap_or("");
        if command.is_empty() {
            return Ok(());
        }

        keywords.push(Keyword::Mmenonic {
            name: command.to_string(),
            line_number,
        });

        while let Some(word) = args.pop_front() {
            match word_type(word, line_number) {
                Ok(Keyword::Mmenonic { name, line_number }) => {
                    return Err(LexerError::CommandAfterCommand {
                        command_name: name,
                        line_number,
                    })
                }
                Ok(keyword) => keywords.push(keyword),
                Err(err) => return Err(err),
            };
        }
    }
    // ends with : -> label
    if let Some(label) = line.strip_suffix(':') {
        keywords.push(Keyword::Label {
            name: label.to_string(),
            line_number,
        });
    }

    Ok(())
}

fn word_type(word: &str, line_number: u16) -> Result<Keyword, LexerError> {
    // register address
    if let Some(register_identifier) = word.strip_prefix('%') {
        return Ok(Keyword::RegisterAddress {
            name: String::from(register_identifier),
            line_number,
        });
    }

    // constant
    // e.g.: 0xa7, 173, 0b0011010
    if let Some(parsed) = if let Some(signed_hex_word) = word.strip_prefix("-0x") {
        Some((signed_hex_word, 16, true))
    } else if let Some(hex_word) = word.strip_prefix("0x") {
        Some((hex_word, 16, false))
    } else if let Some(signed_binary_word) = word.strip_prefix("-0b") {
        Some((signed_binary_word, 2, true))
    } else if let Some(binary_word) = word.strip_prefix("0b") {
        Some((binary_word, 2, false))
    } else if word
        .chars()
        .next()
        .and_then(|first_char| {
            if first_char.is_ascii_digit() {
                Some(())
            } else {
                None
            }
        })
        .is_some()
    {
        Some((word, 10, false))
    } else if let Some(signed_dec_word) = word.strip_prefix('-') {
        if signed_dec_word
            .chars()
            .next()
            .and_then(|first_char| {
                if first_char.is_ascii_digit() {
                    Some(())
                } else {
                    None
                }
            })
            .is_some()
        {
            Some((signed_dec_word, 10, true))
        } else {
            None
        }
    } else {
        None
    }
    .and_then(|(word, radix, sign)| {
        u16::from_str_radix(word, radix)
            .map(|num| if sign { num.wrapping_neg() } else { num })
            .ok()
    }) {
        return Ok(Keyword::Constant {
            value: parsed,
            line_number,
            origin: String::from(word),
        });
    }

    // label
    if word.chars().all(|c| c.is_ascii_alphanumeric() || c == '_') {
        return Ok(Keyword::Label {
            name: String::from(word),
            line_number,
        });
    }

    Err(LexerError::InvalidIdentifier {
        actual: String::from(word),
        line_number,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn lex_all_instructions() {
        let expected = vec![
            Keyword::mmenonic("ldc", 0),
            Keyword::register_address("reg0", 0),
            Keyword::constant("0x00", 0, 0),
            Keyword::mmenonic("ldc", 1),
            Keyword::register_address("reg7", 1),
            Keyword::constant("-0x01", 1u16.wrapping_neg(), 1),
            Keyword::mmenonic("ldc", 2),
            Keyword::register_address("regA", 2),
            Keyword::constant("42", 42, 2),
            Keyword::mmenonic("ldc", 3),
            Keyword::register_address("regH", 3),
            Keyword::constant("-1337", 1337u16.wrapping_neg(), 3),
            Keyword::mmenonic("add", 4),
            Keyword::register_address("reg0", 4),
            Keyword::register_address("reg1", 4),
            Keyword::register_address("reg2", 4),
            Keyword::mmenonic("add3", 5),
            Keyword::register_address("reg3", 5),
            Keyword::register_address("reg4", 5),
            Keyword::register_address("reg5", 5),
            Keyword::register_address("reg6", 5),
            Keyword::mmenonic("addc", 6),
            Keyword::register_address("regB", 6),
            Keyword::register_address("regC", 6),
            Keyword::register_address("regD", 6),
            Keyword::mmenonic("sub", 7),
            Keyword::register_address("regE", 7),
            Keyword::register_address("regF", 7),
            Keyword::register_address("regG", 7),
            Keyword::mmenonic("subc", 8),
            Keyword::register_address("reg0", 8),
            Keyword::register_address("reg0", 8),
            Keyword::register_address("reg0", 8),
            Keyword::mmenonic("inc", 9),
            Keyword::register_address("reg0", 9),
            Keyword::mmenonic("dec", 10),
            Keyword::register_address("reg0", 10),
            Keyword::mmenonic("mul", 11),
            Keyword::register_address("reg0", 11),
            Keyword::register_address("reg0", 11),
            Keyword::register_address("reg0", 11),
            Keyword::mmenonic("and", 12),
            Keyword::register_address("reg0", 12),
            Keyword::register_address("reg0", 12),
            Keyword::register_address("reg0", 12),
            Keyword::mmenonic("or", 13),
            Keyword::register_address("reg0", 13),
            Keyword::register_address("reg0", 13),
            Keyword::register_address("reg0", 13),
            Keyword::mmenonic("not", 14),
            Keyword::register_address("reg0", 14),
            Keyword::register_address("reg0", 14),
            Keyword::mmenonic("neg", 15),
            Keyword::register_address("reg0", 15),
            Keyword::register_address("reg0", 15),
            Keyword::mmenonic("xor", 16),
            Keyword::register_address("reg0", 16),
            Keyword::register_address("reg0", 16),
            Keyword::register_address("reg0", 16),
            Keyword::mmenonic("xnor", 17),
            Keyword::register_address("reg0", 17),
            Keyword::register_address("reg0", 17),
            Keyword::register_address("reg0", 17),
            Keyword::mmenonic("shl", 18),
            Keyword::register_address("reg0", 18),
            Keyword::register_address("reg0", 18),
            Keyword::register_address("reg0", 18),
            Keyword::mmenonic("shr", 19),
            Keyword::register_address("reg0", 19),
            Keyword::register_address("reg0", 19),
            Keyword::register_address("reg0", 19),
            Keyword::mmenonic("tst", 20),
            Keyword::register_address("reg0", 20),
            Keyword::register_address("reg0", 20),
            Keyword::mmenonic("mov", 21),
            Keyword::register_address("reg0", 21),
            Keyword::register_address("reg0", 21),
            Keyword::mmenonic("jmp", 22),
            Keyword::register_address("reg0", 22),
            Keyword::mmenonic("jz", 23),
            Keyword::register_address("reg0", 23),
            Keyword::mmenonic("jnz", 24),
            Keyword::register_address("reg0", 24),
            Keyword::mmenonic("jc", 25),
            Keyword::register_address("reg0", 25),
            Keyword::mmenonic("jrcon", 26),
            Keyword::constant("2047", 2047, 26),
            Keyword::mmenonic("jr", 27),
            Keyword::constant("-2047", 2047u16.wrapping_neg(), 27),
            Keyword::label("jump_0", 28),
            Keyword::mmenonic("jzr", 29),
            Keyword::label("jump_0", 29),
            Keyword::mmenonic("jnzr", 30),
            Keyword::constant("5", 5, 30),
            Keyword::mmenonic("jcr", 31),
            Keyword::constant("5", 5, 31),
            Keyword::mmenonic("st", 32),
            Keyword::register_address("reg0", 32),
            Keyword::register_address("reg1", 32),
            Keyword::mmenonic("ld", 33),
            Keyword::register_address("reg5", 33),
            Keyword::register_address("reg4", 33),
            Keyword::mmenonic("nop", 34),
            Keyword::mmenonic("hlt", 35),
        ];

        let found = lexer(Path::new("tests/all_instructions.s")).unwrap();

        for (expected_keyword, found_keyword) in expected.iter().zip(found.iter()) {
            assert_eq!(expected_keyword, found_keyword);
        }
    }

    #[test]
    fn whitespace() {
        let expected = vec![
            Keyword::mmenonic("ldc", 0),
            Keyword::register_address("reg0", 0),
            Keyword::constant("0x4", 4, 0),
            Keyword::label("loop", 1),
            Keyword::mmenonic("hlt", 2),
        ];

        let found = lexer(Path::new("tests/whitespace.s")).unwrap();
        for (expected_keyword, found_keyword) in expected.iter().zip(found.iter()) {
            assert_eq!(expected_keyword, found_keyword);
        }
    }

    #[test]
    fn comments() {
        let expected = vec![
            Keyword::mmenonic("ldc", 0),
            Keyword::register_address("reg0", 0),
            Keyword::constant("0", 0, 0),
            Keyword::mmenonic("hlt", 1),
        ];

        let found = lexer(Path::new("tests/comments.s")).unwrap();
        for (expected_keyword, found_keyword) in expected.iter().zip(found.iter()) {
            assert_eq!(expected_keyword, found_keyword);
        }
    }
}
